clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2039,24 +2067,34 @@ bool Util::isSyclSpecConstantType(const QualType &Ty) {
 }
 
 bool Util::isSyclType(const QualType &Ty, StringRef Name, bool Tmpl) {
+  if (const CXXRecordDecl *RecTy = Ty->getAsCXXRecordDecl())
+    return isSyclType(RecTy, Name, Tmpl);
+  return false;
+}
+
+bool Util::isSyclType(const CXXRecordDecl *RecTy, StringRef Name, bool Tmpl) {
   Decl::Kind ClassDeclKind =
       Tmpl ? Decl::Kind::ClassTemplateSpecialization : Decl::Kind::CXXRecord;
   std::array<DeclContextDesc, 3> Scopes = {
       Util::DeclContextDesc{clang::Decl::Kind::Namespace, "cl"},
       Util::DeclContextDesc{clang::Decl::Kind::Namespace, "sycl"},
       Util::DeclContextDesc{ClassDeclKind, Name}};
-  return matchQualifiedTypeName(Ty, Scopes);
+  return matchQualifiedTypeName(RecTy, Scopes);
 }
 
 bool Util::matchQualifiedTypeName(const QualType &Ty,
                                   ArrayRef<Util::DeclContextDesc> Scopes) {
+  if (const CXXRecordDecl *RecTy = Ty->getAsCXXRecordDecl())
```

Comment:
Instead of doing this 3 line pattern everywhere, make this whole body be:

    return matchQualifiedTypeName(Ty->getAsCXXRecordDecl(), Scopes);

Then, in matchQualifiedType name, leave the 
    if (!RecTy) return false; 

part.